pkg/reconciler/v1alpha1/pipeline/resources/dag.go

```diff
@@ -84,13 +95,99 @@ func getVisitedPath(path []string) string {
 	return strings.Join(path, " -> ")
 }
 
-//GetPreviousTasks return all the previous tasks for a PipelineTask in the DAG
-func (g *DAG) GetPreviousTasks(pt string) []v1alpha1.PipelineTask {
-	v, ok := g.Nodes[pt]
+// GetSchedulable returns a list of PipelineTask that can be scheduled,
+// given a list of successfully finished doneTasks. It returns task which have
+// all dependecies marked as done, and thus can be scheduled. If the
+// specified doneTasks are invalid (i.e. if it is indicated that a Task is
+// done, but the previous Tasks are not done), an error is returned.
+func (g *DAG) GetSchedulable(doneTasks ...string) ([]v1alpha1.PipelineTask, error) {
+	roots := g.getRoots()
+	tm := toMap(doneTasks...)
+	d := []v1alpha1.PipelineTask{}
+
+	visited := map[string]struct{}{}
+	for _, root := range roots {
+		schedulable := findSchedulable(root, visited, tm)
+		d = append(d, schedulable...)
+	}
+
+	visitedNames := make([]string, len(visited))
+	for v := range visited {
+		visitedNames = append(visitedNames, v)
+	}
+
+	notVisited := list.DiffLeft(doneTasks, visitedNames)
+	if len(notVisited) > 0 {
+		return []v1alpha1.PipelineTask{}, fmt.Errorf("invalid list of done tasks; some tasks were indicated completed without ancestors being done: %v", notVisited)
+	}
+
+	return d, nil
+}
+
+func (g *DAG) getRoots() []*Node {
+	n := []*Node{}
+	for _, node := range g.Nodes {
+		if len(node.Prev) == 0 {
+			n = append(n, node)
+		}
+	}
+	return n
+}
+
+func findSchedulable(n *Node, visited map[string]struct{}, doneTasks map[string]struct{}) []v1alpha1.PipelineTask {
+	if _, ok := visited[n.Task.Name]; ok {
+		return []v1alpha1.PipelineTask{}
+	}
+	visited[n.Task.Name] = struct{}{}
+	if _, ok := doneTasks[n.Task.Name]; ok {
+		schedulable := []v1alpha1.PipelineTask{}
+		// This one is done! Take note of it and look at the next candidate
+		for _, next := range n.Next {
+			if _, ok := visited[next.Task.Name]; !ok {
+				schedulable = append(schedulable, findSchedulable(next, visited, doneTasks)...)
+			}
+		}
+		return schedulable
+	}
+	// This one isn't done! Return it if it's schedulable
+	if isSchedulable(doneTasks, n.Prev) {
+		return []v1alpha1.PipelineTask{n.Task}
+	}
+	// This one isn't done, but it also isn't ready to schedule
+	return []v1alpha1.PipelineTask{}
+}
+
+func isSchedulable(doneTasks map[string]struct{}, prevs []*Node) bool {
+	if len(prevs) == 0 {
+		return true
+	}
+	collected := []string{}
+	for _, n := range prevs {
+		if _, ok := doneTasks[n.Task.Name]; ok {
+			collected = append(collected, n.Task.Name)
+		}
+	}
+	return len(collected) == len(prevs)
+}
+
+func toMap(t ...string) map[string]struct{} {
+	m := make(map[string]struct{}, len(t))
+	for _, s := range t {
+		m[s] = struct{}{}
+	}
+	return m
+}
+
+func addLink(pt v1alpha1.PipelineTask, previousTask string, d *DAG) error {
+	prev, ok := d.Nodes[previousTask]
```

Comment:
@shashwathi it seems, it requires the pointer because `linkPipelineTasks` has the pointer as a receiver… That said, `linkPipelineTasks` doesn't need or use `DAG` anyway, so we could refactor this a bit, looking into it :eyes: